### PR TITLE
[AMORO-2037]Remove dependency on `org.openjdk.jol:jol-core`

### LINKED ISSUE #2037: [Subtask]: Remove dependency on `org.openjdk.jol:jol-core`
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### Description

As [org.openjdk.jol:jol-corea](https://mvnrepository.com/artifact/org.openjdk.jol/jol-core/0.17) is under `The GNU General Public License `which is conflict with `Apache Licence 2.0`, we need to remove our dependency on it to better comply with the specifications of the `Apache License 2.0`.

### Parent issue

#2012

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

Close #2037.

## Brief change log

- Remove dependency on `org.openjdk.jol:jol-core`
- Add `org.apache.lucene:lucene-core` dependency for module core
- Use `org.apache.lucene.util.RamUsageEstimator` to estimate the memory size of object for spillable map

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by zhoujinsong** [comment]
I have tested the results of calculating object memory size using `org.apache.lucene.util.RamUsageEstimator` and `org.openjdk.jol.info.GraphLayout` in my local environment and found:

- The results of calculating for primitive types and collection types are the same.
- The results of calculating for simple objects (objects that do not contain references to other objects) are the same.
- For complex objects, `org.apache.lucene.util.RamUsageEstimator` is inaccurate as it does not recursively calculate the size of referenced objects.
- `org.apache.lucene.util.RamUsageEstimator` has higher execution performance compared to `org.openjdk.jol.info.GraphLayout`, with the former being 20 times faster than the latter.

Because the current `SpillableMap` scenario only requires the calculation of structLike and simple objects, and the calculation process of structLike has been transformed into an array of simple objects, it is feasible to replace `org.openjdk.jol.info.GraphLayout` with `org.apache.lucene.util.RamUsageEstimator` in the current implementation and achieve better performance.

However, it should be noted that `org.apache.lucene.util.RamUsageEstimator` is not a common solution for calculating the size of Java objects.

**Comment by zhoujinsong** [comment]
Additionally, when we downgrade the version of Lucene to 4.0.0, `org.apache.lucene.util.RamUsageEstimator` can accurately calculate the memory usage of objects, but it has compatibility issues with higher versions of JDK and has been abandoned by the Lucene project.

File: trino/pom.xml [hunk 1]
```diff
@@ -353,6 +353,7 @@
         <dependency>
             <groupId>org.openjdk.jol</groupId>
             <artifactId>jol-core</artifactId>
+            <version>${jol.version}</version>
```
Comment (wangtaohz) [inline trino/pom.xml:356]: Does trino moudle still need this dependency `org.openjdk.jol`?

If trino is the only module that must depend on `org.openjdk.jol`, maybe we should declare `jol.version` in the properties of `trino/pom.xml`.

**Comment by zhoujinsong** [comment]
> I noticed that we are using Lucene with `<lucene.version>8.11.2</lucene.version>`. Does it imply that there won't be these compatibility issues with higher versions of JDK?

I think so, the compatibility issues of Lucene version 4.0.0 are because of the use of reflection and it had be removed in higher versions.

File: trino/pom.xml (hunk 1 above)
Comment (zhoujinsong) [reply trino/pom.xml:357]: Yes, I have moved it.

**Comment by shidayang** [comment]
> I have tested the results of calculating object memory size using `org.apache.lucene.util.RamUsageEstimator` and `org.openjdk.jol.info.GraphLayout` in my local environment and found:
> 
> * The results of calculating for primitive types and collection types are the same.
> * The results of calculating for simple objects (objects that do not contain references to other objects) are the same.
> * For complex objects, `org.apache.lucene.util.RamUsageEstimator` is inaccurate as it does not recursively calculate the size of referenced objects.
> * `org.apache.lucene.util.RamUsageEstimator` has higher execution performance compared to `org.openjdk.jol.info.GraphLayout`, with the former being 20 times faster than the latter.
> 
> Because the current `SpillableMap` scenario only requires the calculation of structLike and simple objects, and the calculation process of structLike has been transformed into an array of simple objects, it is feasible to replace `org.openjdk.jol.info.GraphLayout` with `org.apache.lucene.util.RamUsageEstimator` in the current implementation and achieve better performance.
> 
> However, it should be noted that `org.apache.lucene.util.RamUsageEstimator` is not a common solution for calculating the size of Java objects.

It seem RamUsageEstimator only can calculator know type, other type is always 256

**Review by shidayang — APPROVED** [review]

**Review by wangtaohz — APPROVED** [review]
LGTM

**Pull request merged**